lib/msal-common/src/network/ThrottlingUtils.ts

```diff
@@ -0,0 +1,114 @@
+ * Licensed under the MIT License.
+ */
+
+import { NetworkResponse } from "./NetworkManager";
+import { ServerAuthorizationTokenResponse } from "../server/ServerAuthorizationTokenResponse";
+import { Constants, HeaderNames } from "../utils/Constants";
+import { CacheManager } from "../cache/CacheManager";
+import { StringUtils } from "../utils/StringUtils";
+import { ServerError } from "../error/ServerError";
+import { RequestThumbprint } from "./RequestThumbprint";
+import { RequestThumbprintValue } from "./RequestThumbprintValue";
+
+export class ThrottlingUtils {
+
+    /**
+     * Prepares a RequestThumbprint to be stored as a key.
+     * @param thumbprint
+     */
+    static generateThrottlingStorageKey(thumbprint: RequestThumbprint): string {
+        return `${Constants.THROTTLE_PREFIX}.${JSON.stringify(thumbprint)}`;
+    }
+
+    /**
+     * Performs necessary throttling checks before a network request.
+     * @param cacheManager
+     * @param thumbprint 
+     */
+    static preProcess(cacheManager: CacheManager, thumbprint: RequestThumbprint): void {
+        const key = ThrottlingUtils.generateThrottlingStorageKey(thumbprint);
+        const storageValue = cacheManager.getItem(key) as string;
+
+        if (storageValue) {
+            const parsedValue = StringUtils.jsonParseHelper<RequestThumbprintValue>(storageValue);
+
+            if (parsedValue) {
+                if (parsedValue.throttleTime < Date.now()) {
+                    cacheManager.removeItem(key);
+                    return;
+                }
+                throw new ServerError(parsedValue.errorCodes.join(" "), parsedValue.errorMessage, parsedValue.subError);
+            }
+        }
+    }
+
+    /**
+     * Performs necessary throttling checks after a network request.
+     * @param cacheManager
+     * @param thumbprint 
+     * @param response
+     */    
+    static postProcess(cacheManager: CacheManager, thumbprint: RequestThumbprint, response: NetworkResponse<ServerAuthorizationTokenResponse>): void {
+        if (ThrottlingUtils.checkResponseStatus(response) || ThrottlingUtils.checkResponseForRetryAfter(response)) {
+            const thumbprintValue: RequestThumbprintValue = {
+                throttleTime: ThrottlingUtils.calculateThrottleTime(parseInt(response.headers.get(HeaderNames.RETRY_AFTER))),
+                error: response.body.error,
+                errorCodes: response.body.error_codes,
+                errorMessage: response.body.error_description,
+                subError: response.body.suberror
+            };
+            cacheManager.setItem(
+                ThrottlingUtils.generateThrottlingStorageKey(thumbprint),
+                JSON.stringify(thumbprintValue)
+            );
+        }
+    }
+
+    /**
+     * Checks a NetworkResponse object's status codes against 429 or 5xx
+     * @param response
+     */
+    static checkResponseStatus(response: NetworkResponse<ServerAuthorizationTokenResponse>): boolean {
```

Comment:
FYI I implemented this function slightly differently in Telemetry and included it in the NetworkManager class. Take a look at that PR and see if that might overlap with this here

Reply:
They're slightly different use cases, though both check against `status`. I'm going to leave this as-is for now.